pallets/providers/src/lib.rs

```diff
@@ -0,0 +1,367 @@
+            + scale_info::TypeInfo
+            + MaxEncodedLen;
+
+        /// The minimum amount that an account has to deposit to become a storage provider.
+        #[pallet::constant]
+        type SpMinDeposit: Get<BalanceOf<Self>>;
+
+        /// The amount that a BSP receives as allocation of storage capacity when it deposits SpMinDeposit.
+        #[pallet::constant]
+        type SpMinCapacity: Get<StorageData<Self>>;
+
+        /// The slope of the collateral vs storage capacity curve. In other terms, how many tokens a Storage Provider should add as collateral to increase its storage capacity in one unit of StorageData.
+        #[pallet::constant]
+        type DepositPerData: Get<BalanceOf<Self>>;
+
+        /// The maximum amount of BSPs that can exist.
+        #[pallet::constant]
+        type MaxBsps: Get<u32>;
+
+        /// The maximum size of a multiaddress.
+        #[pallet::constant]
+        type MaxMultiAddressSize: Get<u32>;
+
+        /// The maximum number of protocols the MSP can support (at least within the runtime).
+        #[pallet::constant]
+        type MaxProtocols: Get<u32>;
+    }
+
+    #[pallet::pallet]
+    pub struct Pallet<T>(_);
+
+    // Storage:
+
+    /// The mapping from an account id to a main storage provider. Returns `None` if the account isnt a main storage provider.
+    #[pallet::storage]
+    pub type Msps<T: Config> =
+        StorageMap<_, Blake2_128Concat, T::AccountId, MainStorageProvider<T>>;
+
+    /// The mapping from an account id to a backup storage provider. Returns `None` if the account isnt a backup storage provider.
+    #[pallet::storage]
+    pub type Bsps<T: Config> =
+        StorageMap<_, Blake2_128Concat, T::AccountId, BackupStorageProvider<T>>;
+
+    /// The amount of storage providers (MSPs and BSPs) that are currently registered in the runtime.
+    #[pallet::storage]
+    pub type UserCount<T: Config> = StorageValue<_, T::UserCount, ValueQuery>;
+
+    /// The total amount of storage capacity all BSPs have. Remember redundancy!
+    #[pallet::storage]
+    #[pallet::getter(fn total_bsps_capacity)] // TODO: remove this and add an explicit getter
+    pub type TotalBspsCapacity<T: Config> = StorageValue<_, StorageData<T>>;
+
+    /// A vector of account IDs that are BSPs (to draw from for proofs)
+    #[pallet::storage]
+    pub type BspsVec<T: Config> = StorageValue<_, BoundedVec<T::AccountId, T::MaxBsps>, ValueQuery>;
+
+    // Events & Errors:
+
+    /// The events that can be emitted by this pallet
+    #[pallet::event]
+    #[pallet::generate_deposit(pub(super) fn deposit_event)]
+    pub enum Event<T: Config> {
+        /// Event emitted when a Main Storage Provider has signed up successfully. Provides information about
+        /// that MSP's account id, the total data it can store according to its stake, its multiaddress, and its value proposition.
+        MspSignUpSuccess {
+            who: T::AccountId,
+            info: MainStorageProvider<T>,
```

Comment:
I'd consider expanding the `info` field for this event. This will make the event more easily readable when checked on-chain. No biggie though.